File: src/valueVerification.ts
import type { VerificationScalar } from "./types.js";

export function canonicalVerificationDisplayForExpected(v: VerificationScalar): string {
  if (v === null) return "null";
  if (typeof v === "boolean") return v ? "true" : "false";
  if (typeof v === "number") {
    if (Number.isNaN(v)) return "NaN";
    if (v === Infinity) return "Infinity";
    if (v === -Infinity) return "-Infinity";
    return String(v);
  }
  return JSON.stringify(v);
}

export function canonicalVerificationDisplayForActual(actual: unknown): string {
  if (actual === null || actual === undefined) return "null";
  if (typeof actual === "boolean") return actual ? "true" : "false";
  if (typeof actual === "number") {
    if (Number.isNaN(actual)) return "NaN";
    if (actual === Infinity) return "Infinity";
    if (actual === -Infinity) return "-Infinity";
    return String(actual);
  }
  if (typeof actual === "string") return JSON.stringify(actual);
  if (typeof actual === "bigint") return JSON.stringify(actual.toString());
  if (actual instanceof Date && !Number.isNaN(actual.getTime())) {
    return JSON.stringify(actual.toISOString());
  }
  return JSON.stringify(String(actual));
}

export type VerificationCompareResult =
  | { ok: true }
  | { ok: false; expected: string; actual: string };

function fail(expected: VerificationScalar, actual: unknown): VerificationCompareResult {
  return {
    ok: false,
    expected: canonicalVerificationDisplayForExpected(expected),
    actual: canonicalVerificationDisplayForActual(actual),
  };
}

export function verificationScalarsEqual(
  expected: VerificationScalar,
  actual: unknown,
): VerificationCompareResult {
  if (expected === null) {
    if (actual === null || actual === undefined) return { ok: true };
    return fail(expected, actual);
  }

  if (actual === null || actual === undefined) {
    return fail(expected, actual);
  }

  if (typeof expected === "boolean") {
    const ok =
      (typeof actual === "boolean" && actual === expected) ||
      (typeof actual === "number" &&
        Number.isFinite(actual) &&
        ((expected === true && actual === 1) || (expected === false && actual === 0)));
    return ok ? { ok: true } : fail(expected, actual);
  }

  if (typeof expected === "number") {
    if (!Number.isFinite(expected)) {
      return fail(expected, actual);
    }

    if (typeof actual === "number" && Number.isFinite(actual) && actual === expected) {
      return { ok: true };
    }

    if (
      typeof actual === "bigint" &&
      Number.isInteger(expected) &&
      expected >= Number.MIN_SAFE_INTEGER &&
      expected <= Number.MAX_SAFE_INTEGER &&
      actual === BigInt(expected)
    ) {
      return { ok: true };
    }

    if (typeof actual === "string") {
      const t = actual.trim();
      try {
        const parsed = JSON.parse(t) as unknown;
        if (
          typeof parsed === "number" &&
          Number.isFinite(parsed) &&
          parsed === expected &&
          JSON.stringify(parsed) === JSON.stringify(expected)
        ) {
          return { ok: true };
        }
      } catch {
        /* no match */
      }
    }

    return fail(expected, actual);
  }

  if (typeof expected === "string") {
    const e = expected.trim();

    if (typeof actual === "string" && actual.trim() === e) {
      return { ok: true };
    }

    if (typeof actual === "number" && Number.isFinite(actual) && e === JSON.stringify(actual)) {
      return { ok: true };
    }

    if (typeof actual === "boolean" && e === JSON.stringify(actual)) {
      return { ok: true };
    }

    if (actual instanceof Date && !Number.isNaN(actual.getTime()) && e === actual.toISOString()) {
      return { ok: true };
    }

    return fail(expected, actual);
  }

  return fail(expected, actual);
}

File: python/src/agentskeptic/kernel/value_verification.py
from __future__ import annotations

import json
import math
from datetime import datetime
from typing import Any, Literal, TypedDict, Union

VerificationScalar = Union[str, int, float, bool, None]


class CmpOk(TypedDict):
    ok: Literal[True]


class CmpFail(TypedDict):
    ok: Literal[False]
    expected: str
    actual: str


Cmp = CmpOk | CmpFail


def canonical_verification_display_for_expected(v: VerificationScalar) -> str:
    if v is None:
        return "null"
    if isinstance(v, bool):
        return "true" if v else "false"
    if isinstance(v, float):
        if math.isnan(v):
            return "NaN"
        if v == float("inf"):
            return "Infinity"
        if v == float("-inf"):
            return "-Infinity"
        return str(v)
    if isinstance(v, int):
        return str(v)
    return json.dumps(v)


def canonical_verification_display_for_actual(actual: Any) -> str:
    if actual is None:
        return "null"
    if isinstance(actual, bool):
        return "true" if actual else "false"
    if isinstance(actual, float):
        if math.isnan(actual):
            return "NaN"
        if actual == float("inf"):
            return "Infinity"
        if actual == float("-inf"):
            return "-Infinity"
        return str(actual)
    if isinstance(actual, int):
        return str(actual)
    if isinstance(actual, str):
        return json.dumps(actual)
    if isinstance(actual, bytes):
        return json.dumps(actual.decode("utf-8", errors="replace"))
    if isinstance(actual, datetime):
        if math.isnan(actual.timestamp()):
            return json.dumps(None)
        return json.dumps(actual.isoformat())
    return json.dumps(str(actual))


def verification_scalars_equal(expected: VerificationScalar, actual: Any) -> Cmp:
    def fail() -> CmpFail:
        return {
            "ok": False,
            "expected": canonical_verification_display_for_expected(expected),
            "actual": canonical_verification_display_for_actual(actual),
        }

    if expected is None:
        if actual is None:
            return {"ok": True}
        return fail()

    if actual is None:
        return fail()

    if isinstance(expected, bool):
        ok = (isinstance(actual, bool) and actual is expected) or (
            isinstance(actual, (int, float))
            and math.isfinite(float(actual))
            and ((expected is True and float(actual) == 1.0) or (expected is False and float(actual) == 0.0))
        )
        return {"ok": True} if ok else fail()

    if isinstance(expected, (int, float)):
        if isinstance(expected, float) and not math.isfinite(expected):
            return fail()
        exp = float(expected) if isinstance(expected, int) else expected
        if isinstance(actual, (int, float)) and math.isfinite(float(actual)) and float(actual) == float(exp):
            return {"ok": True}
        if isinstance(actual, int) and isinstance(expected, int) and actual == expected:
            return {"ok": True}
        if isinstance(actual, str):
            try:
                parsed = json.loads(actual.strip())
                if isinstance(parsed, (int, float)) and float(parsed) == float(exp):
                    return {"ok": True}
            except (json.JSONDecodeError, TypeError, ValueError):
                pass
        return fail()

    if isinstance(expected, str):
        e = expected.strip()
        if isinstance(actual, str) and actual.strip() == e:
            return {"ok": True}
        if isinstance(actual, (int, float)) and math.isfinite(float(actual)) and e == json.dumps(actual):
            return {"ok": True}
        if isinstance(actual, bool) and e == json.dumps(actual):
            return {"ok": True}
        if isinstance(actual, datetime) and e == actual.isoformat():
            return {"ok": True}
        return fail()

    return fail()
